File: vignettes/04_streaming/csharp/04_streaming.cs
// Streaming Responses — Real-time token delivery (C#)
//
// This sample demonstrates how to stream responses from an agent,
// receiving tokens as they are generated rather than waiting for the
// full response. It mirrors the Julia vignette 04_streaming.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// Create an agent backed by the local Ollama instance.
AIAgent agent = new OllamaApiClient(new Uri(endpoint), modelName)
    .AsAIAgent(
        instructions: "You are a helpful assistant. Provide detailed answers.",
        name: "StreamingAgent");

// ── Non-streaming: wait for the full response ────────────────────────────
Console.WriteLine("=== Non-streaming ===");
Console.WriteLine(await agent.RunAsync("Explain the water cycle in one paragraph."));
Console.WriteLine();

// ── Streaming: receive tokens as they arrive ─────────────────────────────
Console.WriteLine("=== Streaming ===");
Console.Write("Agent: ");
await foreach (var update in agent.RunStreamingAsync(
    "Explain the water cycle in one paragraph."))
{
    Console.Write(update);
}
Console.WriteLine();
Console.WriteLine();

// ── Streaming with token inspection ──────────────────────────────────────
Console.WriteLine("=== Token inspection ===");
var tokenCount = 0;
await foreach (var update in agent.RunStreamingAsync("What is 2 + 2?"))
{
    var text = update.ToString();
    if (!string.IsNullOrEmpty(text))
    {
        tokenCount++;
        Console.WriteLine($"  Fragment {tokenCount}: \"{text}\"");
    }
}
Console.WriteLine($"Total fragments: {tokenCount}");

File: vignettes/04_streaming/python/04_streaming.py
"""
Streaming Responses — Real-time token delivery (Python)

This sample demonstrates how to stream responses from an agent,
receiving tokens as they are generated rather than waiting for the
full response. It mirrors the Julia vignette 04_streaming.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio

from agent_framework.ollama import OllamaChatClient


async def main() -> None:
    # Create a chat client pointing at the local Ollama instance.
    client = OllamaChatClient(
        host="http://localhost:11434",
        model_id="qwen3:8b",
    )

    # Build an agent with instructions.
    agent = client.as_agent(
        name="StreamingAgent",
        instructions="You are a helpful assistant. Provide detailed answers.",
    )

    # ── Non-streaming: wait for the full response ────────────────────────
    print("=== Non-streaming ===")
    result = await agent.run("Explain the water cycle in one paragraph.")
    print(f"Agent: {result}")
    print()

    # ── Streaming: receive tokens as they arrive ─────────────────────────
    print("=== Streaming ===")
    print("Agent: ", end="", flush=True)
    async for chunk in agent.run(
        "Explain the water cycle in one paragraph.", stream=True
    ):
        if chunk.text:
            print(chunk.text, end="", flush=True)
    print()
    print()

    # ── Streaming with token inspection ──────────────────────────────────
    print("=== Token inspection ===")
    token_count = 0
    async for chunk in agent.run("What is 2 + 2?", stream=True):
        if chunk.text:
            token_count += 1
            print(f"  Fragment {token_count}: {chunk.text!r}")
    print(f"Total fragments: {token_count}")


if __name__ == "__main__":
    asyncio.run(main())
